asmcomp/cmm.ml

```diff
@@ -32,6 +32,39 @@ let size_component = function
   | Int -> Arch.size_int
   | Float -> Arch.size_float
 
+let lub_component comp1 comp2 =
+  match comp1, comp2 with
+  | Int, Int -> Int
+  | Int, Addr -> Addr
+  | Int, Val -> Val
+  | Addr, Int -> Addr
+  | Addr, Addr -> Addr
+  | Addr, Val -> Val
```

Comment:
"Addr" denotes a derived pointer that might point in the middle of a heap block.  The GC is unable to treat such derived pointers correctly: it cannot follow them and it cannot update them after moving the underlying object.  This is why it is a compile-time fatal error to have a variable of type Addr live across a GC or call point.

Now, if you have a conditional where x : Addr in one branch and x : Val in the other, x may be a derived pointer and therefore must be given type Addr after the join point.  If it dies before the next GC / call point, everything will be fine.  Otherwise it is a serious code generation error and the compiler should bomb.

In summary: Addr, Val -> Addr and symmetrically.